### PR TITLE
Media Type enforcement

### LINKED ISSUE #702: Handling of urls which don't have file extension
Currently deno doesn't seem handling a url which doesn't have a file extension. However browsers seem working with such urls.

```html
<script type="module">
  import React from 'https://dev.jspm.io/react';
</script>
```

I think deno should handle those urls as JavaScript because that's more browser compatible and because it allows deno to work with https://dev.jspm.io. If deno works with jspm.io, I think it enables a lot of more meaningful experiments which depends on npm modules.

### PR DESCRIPTION
**Pull request by kitsonk**
Fixes #702 

This PR is a WIP for media type enforcement.

When remote modules are retrieved, the privileged resolves the media type of the file.  First it respects the `Content-Type` header if present.  If no header present, it will utilise the extension.  For local modules the privileged side always uses the extension to determine the media type.

The Deno compiler then respects that media type when providing file type information to TypeScript.

Things not done yet:

- [x] When the code is cached, the media type needs to be cached as well, so when it is retrieved it can be passed on to userland.
- [x] Tests.
- [x] gist returns `text/plain` for files.  At the moment, that assigns it a `MediaType.Unknown` which the compiler assumes is then JavaScript.  What to do with media types that are not "valid" needs to be figured out.

### CODE REVIEW

File: src/deno_dir.rs [hunk 1]
```diff
@@ -104,26 +105,59 @@ impl DenoDir {
     }
   }
 
+  // convert a ContentType string into a enumerated MediaType
+  fn map_content_type(self: &DenoDir, path: &Path, content_type: Option<&str>) -> msg::MediaType {
```
[the remark this answers is not in the document]
Comment (kitsonk) [reply src/deno_dir.rs]: Ok, I will, I just was struggling with how to invoke it.  I will figure that out... I guess it needs to go outside of the `impl`.  It is obvious now that I say it.

**Comment by kitsonk** [comment]
@ry do you have an opinion on how we should resolve _unknown_ media types?  Specifically, gist returns things as `text/plain`.  Currently Rust indicates it is `MediaType.Unknown` but Deno compiler assumes that anything _unknown_ is JavaScript.

A couple solutions I can think of:

- We really follow the `Content-Type` strictly, so if we have a media type that is not _known_ that we strictly pass `Unknown` along and the Deno compiler will throw terminally if an _unknown_ module is resolved.
- If we find a `Content-Type` that we don't recognise, Rust then falls back to looking at the extension.  Currently if we don't get a `Content-Type` or it is a local file, this is how it is done.
- We specifically only look at extension for a couple key media types, like `text/plain`.

Browsers tend to be really really strict about it, so it feels like we should be too.  It does mean that we can't host `.ts` or `.js` files on Gist (or we have to point at other URIs that proxy it and provide the correct media types).

**Comment by ry** [comment]
@kitsonk Looking at the extension should solve it for gists ? I think we should use the extension. Otherwise I have no opinion.

**Comment by kitsonk** [comment]
@ry ok, I think this is now complete and ready for review.

So for the "gist" issue, if the media type is `text/plain` then Deno reverts to looking at the extension.  If it isn't one of the _known_ media types though, privileged will pass `MediaType.Unknown` and the Deno compiler with terminally throw indicating it can't deal with an unknown media type.

The media type that is received for remote modules is cached along side the source code with a `.mime` extension, which is then read back in when a cached remote module is read.  Local modules have their type determined simply by extension.

File: tools/http_server.py [hunk 2]
```diff
@@ -14,9 +14,33 @@
...
+            return "video/mp2t"
+        if ".j1." in path:
+            return "text/javascript"
+        if ".j2." in path:
+            return "application/ecmascript"
+        if ".j3." in path:
+            return "text/ecmascript"
+        if ".j4." in path:
+            return "application/x-javascript"
+        return SimpleHTTPServer.SimpleHTTPRequestHandler.guess_type(self, path)
```
Comment (ry) [inline tools/http_server.py]: Nice.

File: src/http_util.rs [hunk 3]
```diff
@@ -87,7 +97,8 @@ fn test_fetch_sync_string() {
 fn test_fetch_sync_string_with_redirect() {
   // Relies on external http server. See tools/http_server.py
   tokio_util::init(|| {
-    let p = fetch_sync_string("http://127.0.0.1:4546/package.json").unwrap();
+    let (p, _m) =
+      fetch_sync_string("http://127.0.0.1:4546/package.json").unwrap();
     println!("package.json len {}", p.len());
```
Comment (ry) [inline src/http_util.rs]: Should you not test `_m` here?

File: src/http_util.rs [hunk 4]
```diff
@@ -77,7 +86,8 @@ pub fn fetch_sync_string(module_name: &str) -> DenoResult<String> {
 fn test_fetch_sync_string() {
   // Relies on external http server. See tools/http_server.py
   tokio_util::init(|| {
-    let p = fetch_sync_string("http://127.0.0.1:4545/package.json").unwrap();
+    let (p, _m) =
+      fetch_sync_string("http://127.0.0.1:4545/package.json").unwrap();
```
Comment (ry) [inline src/http_util.rs]: Should you not test `_m` here?

File: src/http_util.rs [hunk 5]
```diff
@@ -27,7 +28,7 @@ pub fn get_client() -> Client<Connector, hyper::Body> {
 
 // The CodeFetch message is used to load HTTP javascript resources and expects a
 // synchronous response, this utility method supports that.
-pub fn fetch_sync_string(module_name: &str) -> DenoResult<String> {
+pub fn fetch_sync_string(module_name: &str) -> DenoResult<(String, String)> {
```
Comment (ry) [inline src/http_util.rs]: How about returning `DenoResult<(String, msg::MediaType)>` ?

File: src/deno_dir.rs [hunk 6]
```diff
@@ -646,3 +656,157 @@ fn parse_local_or_remote(p: &str) -> Result<url::Url, url::ParseError> {
...
+  );
+  assert_eq!(
+    map_content_type(Path::new("foo/bar.ts"), Some("text/plain")),
+    msg::MediaType::TypeScript
+  );
+  assert_eq!(
+    map_content_type(Path::new("foo/bar.ts"), Some("foo/bar")),
+    msg::MediaType::Unknown
+  );
+}
```
Comment (ry) [inline src/deno_dir.rs]: Nice

File: src/deno_dir.rs [hunk 7]
```diff
@@ -646,3 +656,157 @@ fn parse_local_or_remote(p: &str) -> Result<url::Url, url::ParseError> {
...
+  );
+  assert_eq!(
+    map_file_extension(Path::new("foo/bar.txt")),
+    msg::MediaType::Unknown
+  );
+  assert_eq!(
+    map_file_extension(Path::new("foo/bar")),
+    msg::MediaType::Unknown
+  );
+}
```
Comment (ry) [inline src/deno_dir.rs]: 👍

File: src/deno_dir.rs [hunk 8]
```diff
@@ -109,21 +110,26 @@ impl DenoDir {
     self: &DenoDir,
     module_name: &str,
     filename: &str,
-  ) -> DenoResult<String> {
+  ) -> DenoResult<(String, msg::MediaType)> {
     let p = Path::new(filename);
+    let mut media_type_filename = filename.to_string();
+    media_type_filename.push_str(".mime");
+    let mt = Path::new(&media_type_filename);
```
Comment (ry) [inline src/deno_dir.rs]: Can you add a comment to this section. Something along the lines of:

```
// We write a special ".mime" file into the .deno/deps directory along side the
// cached file, containing just the mime type.
```

File: js/os.ts [hunk 9]
```diff
@@ -1,11 +1,18 @@
 // Copyright 2018 the Deno authors. All rights reserved. MIT license.
-import { ModuleInfo } from "./types";
 import * as msg from "gen/msg_generated";
 import { assert } from "./util";
 import * as util from "./util";
 import * as flatbuffers from "./flatbuffers";
 import { sendSync } from "./dispatch";
 
+interface ModuleInfo {
```
Comment (ry) [inline js/os.ts]: No longer @internal?

Why is this moved out of types? I guess because it relies on msg.MediaType ?

Maybe this should be in compiler.ts ?

**Comment by ry** [comment]
Let's try to land this before #1039

File: src/http_util.rs (hunk 3 above)
Comment (kitsonk) [reply src/http_util.rs]: doh! forgot when I broke them to actually adjust them

File: src/deno_dir.rs (hunk 8 above)
Comment (kitsonk) [reply src/deno_dir.rs]: 👍

File: js/os.ts (hunk 9 above)
Comment (kitsonk) [reply js/os.ts]: Well it isn't exported anymore so it shouldn't leak.  It was only `@internal` because it was an explicit export of a module, but I forgot to double check that somehow it didn't leak.

I will see if it fits better in `compiler.ts`.

**Comment by kitsonk** [comment]
Hmmm... the appveyor `./tools/format.py` reformatted one of the files, but when I run it locally, I don't get any changes.

**Comment by kitsonk** [comment]
Hmmm... the appveyor `./tools/format.py` reformatted one of the files, but when I run it locally, I don't get any changes.

**Comment by ry** [comment]
@kitsonk Hmm - strange - make sure you're rebased on master and using the rustfmt in third_party.

**Comment by kitsonk** [comment]
@ry I was/am and I was using `./tools/format.py` and validated it used the one in `third_party`.  It could have been any number of things with the way GitHub was performing.  I will double check again in a few hours.

**Comment by kitsonk** [comment]
@ry 😖 😕 

```
$ git log
commit 080431e5c7acd6e558d6cdbdb8c199c622e6899b (HEAD -> media-types, origin/media-types)
Author: Kitson Kelly <me@kitsonkelly.com>
Date:   Mon Oct 22 13:14:27 2018 +1100

    Enforce media types

commit c4bddc4651f583241193b80634a4f22abb02d582 (upstream/master, origin/master, origin/HEAD, master)
Author: ztplz <mysticzt@gmail.com>
Date:   Mon Oct 22 05:42:18 2018 +0800

    Implemente  clone for FetchResponse (#1054)

commit 47c96a61527f56aa6dfc5a8fc44b13a1db4f4da5
Author: Ryan Dahl <ry@tinyclouds.org>
Date:   Sun Oct 21 13:43:47 2018 -0400

    CI should fail when code isn't formatted.

commit ae0dec39dacbfa4279d1d708bf6b7488c230c272
Author: Ryan Dahl <ry@tinyclouds.org>
Date:   Sun Oct 21 16:09:05 2018 -0400
$ ./tools/format.py
... snip ...
third_party/rustfmt/mac/rustfmt --config-path /Users/kkelly/github/deno/tools/rustfmt.toml src/flags.rs src/tokio_write.rs src/version.rssrc/http_util.rs src/deno_dir.rs src/errors.rs src/main.rs src/tokio_util.rs src/fs.rs src/resources.rs src/isolate.rs src/ops.rs src/libdeno.rs
$ git status
On branch media-types
Your branch is up to date with 'origin/media-types'.

nothing to commit, working tree clean
```

**Comment by kitsonk** [comment]
Well, and #1065 which currently includes the commit here passed CI on appveyor.  Seems to be some sort of odd issue while GitHub was 💥.  I suspect restarting the CI would cause the issue to disappear (as well as complete the Travis CI as well).

File: src/http_util.rs (hunk 5 above)
Comment (kitsonk) [reply src/http_util.rs]: I took a look at this and I decided not to change it... within `http_util` it doesn't really have a concept of the enum yet, as that is decided in `deno_dir.rs`.  So a decision here, might not be _right_ when combined with other logic contained in `deno_dir.rs`, so I would like to leave it as `(String, String)`.
